### ISSUE TITLE
Increased augmentation time after Quarkus 3.8.4 upgrade

### ISSUE THREAD

**Issue by cwelch-rh** [keycloak/keycloak#29579]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

dist/quarkus

### Describe the bug

Quarkus augmentation time significantly increased after the upgrade to 3.8.4. It appears to mostly be due to the changes in the ResteasyReactiveJacksonProcessor#handleFieldSecurity method.

### Version

Nightly

### Regression

- [X] The issue is a regression

### Expected behavior

Not so drastic impact on augmentation time. Timing info from 3.8.3:

Without Providers:
2024-05-13 17:55:51,949 TRACE [io.qua.builder] (build-8) Finished step "io.quarkus.resteasy.reactive.jackson.deployment.processor.ResteasyReactiveJacksonProcessor#handleFieldSecurity" in 17 ms

With Providers:
2024-05-13 17:51:30,489 TRACE [io.qua.builder] (build-64) Finished step "io.quarkus.resteasy.reactive.jackson.deployment.processor.ResteasyReactiveJacksonProcessor#handleFieldSecurity" in 19 ms

### Actual behavior

Large jump in augmentation time when using our custom providers. The time increase is almost unnoticeable without providers.

Without Providers:
2024-05-02 19:25:17,643 TRACE [io.qua.builder] (build-28) Finished step "io.quarkus.resteasy.reactive.jackson.deployment.processor.ResteasyReactiveJacksonProcessor#handleFieldSecurity" in 827 ms

With Providers:
2024-05-02 19:19:52,537 TRACE [io.qua.builder] (build-23) Finished step "io.quarkus.resteasy.reactive.jackson.deployment.processor.ResteasyReactiveJacksonProcessor#handleFieldSecurity" in 44565 ms

### How to Reproduce?

Download or build the nightly keycloak quarkus distribution.
Enable TRACE for io.quarkus
Run /bin/kc.sh build

The increase in time can be seen without providers, but it's more dramatic with ours.

### Anything else?

This commit in quarkus introduced a bunch of recursion in related code.
https://github.com/quarkusio/quarkus/commit/e6376f9e5b4623989745079786d9aa0fd6bf6d72

It looks like it's checking return types from REST resource methods, but I'm having trouble tracking down exactly what about our RealmResourceProviders might be causing the increase in build time.

**Comment by mabartos** [comment]
Blocked externally by: https://github.com/quarkusio/quarkus/issues/40780

**Comment by mabartos** [comment]
@michalvavrik Yep, probably better to reference the https://github.com/quarkusio/quarkus/issues/40153 with the 3.8.5 milestone.

Do you know the ETA for Quarkus 3.8.5 release?

**Comment by michalvavrik** [comment]
> Do you know the ETA for Quarkus 3.8.5 release?

I'd expect it this week, not sure if plans didn't change as I was on PTO. Better contact someone participating in the release process as Guillaume.

**Comment by VinodAnandan** [comment]
@gsmet Could you please let us know the ETA for Quarkus 3.8.5 release?

**Comment by VinodAnandan** [comment]
@vmuzikar Could you please consider including the Quarkus 3.8.5 update as part of Keycloak [25.0.0](https://github.com/keycloak/keycloak/milestone/47)? There are multiple security fixes.

**Comment by vmuzikar** [comment]
@VinodAnandan Quarkus 3.8.5 has not been released yet. We'll upgrade to it in a micro release.

**Comment by VinodAnandan** [comment]
I apologise for the misunderstanding. I thought the Keycloak 25 version was also waiting for the Quarkus 3.8.5 release.

It is my understanding that the Quarkus team plans to release version 3.8.5 within the next few days. Releasing a major Keycloak version with known vulnerabilities may not be the best practice, as most large organisations prioritise supply chain security and conduct thorough ingestion scans. If any known vulnerabilities are detected, these releases are typically blocked.

Furthermore, it appears that Quarkus 3.8.x has less priority/attention compared to 3.9.x and 3.10.x versions. Can we reliably depend on the Quarkus 3.8.x versions?

**Comment by gsmet** [comment]
> Furthermore, it appears that Quarkus 3.8.x has less priority/attention compared to 3.9.x and 3.10.x versions. Can we reliably depend on the Quarkus 3.8.x versions?

That's not true.
3.9 and 3.10 are not maintained anymore, only 3.11 is currently maintained.

The lifecycle of 3.8 is slower as we are very careful about what we backport there and we also wait for some bake time in the latest. But it has our full attention.

**Comment by VinodAnandan** [comment]
@gsmet  Thanks for the clarification. I made the above comments based on the data available here: https://mvnrepository.com/artifact/io.quarkus/quarkus-core.

Would you recommend that Keycloak rely on version 3.11.x instead of 3.8.x? Is there any standard wait time for security patches to be backported to 3.8.x?

**Comment by vmuzikar** [comment]
> Releasing a major Keycloak version with known vulnerabilities may not be the best practice

That is true, but AFAIK Keycloak is not vulnerable to any of the CVEs affecting 3.8.4.

CC @abstractj

> Would you recommend that Keycloak rely on version 3.11.x instead of 3.8.x?

Keycloak is sticking with fully supported LTS versions of Quarkus (currently 3.8) which are also corresponding to Red Hat build of Quarkus that is used in Red Hat build of Keycloak.

**Comment by VinodAnandan** [comment]
Thank you all for releasing Quarkus 3.8.5.

Transitive library vulnerabilities, particularly those related to cryptography, and their severity scores can be complex topics. It might be best to discuss these matters separately from this issue, particularly since this issue is related to regression.

> That is true, but AFAIK Keycloak is not vulnerable to any of the CVEs affecting 3.8.4.

Below are some of the vulnerabilities affecting Keycloak due to the use of Quarkus 3.8.4:

https://security.snyk.io/vuln/SNYK-JAVA-ORGBOUNCYCASTLE-6612984
https://security.snyk.io/vuln/SNYK-JAVA-ORGBOUNCYCASTLE-3136316
https://security.snyk.io/vuln/SNYK-JAVA-ORGBOUNCYCASTLE-7159336
https://security.snyk.io/vuln/SNYK-JAVA-ORGBOUNCYCASTLE-6613076
https://security.snyk.io/vuln/SNYK-JAVA-ORGBOUNCYCASTLE-6613079
https://security.snyk.io/vuln/SNYK-JAVA-ORGBOUNCYCASTLE-2841508
https://security.snyk.io/vuln/SNYK-JAVA-ORGBOUNCYCASTLE-1296075
https://security.snyk.io/vuln/SNYK-JAVA-ORGBOUNCYCASTLE-5771339
https://security.snyk.io/vuln/SNYK-JAVA-ORGBOUNCYCASTLE-6084022

Considering that Quarkus 3.8.5 has been released and both the vulnerability and this particular regression issue will be addressed by updating to Quarkus 3.8.5, is it okay to include this issue in the Keycloak 25.0.0 milestone?

**Comment by ahus1** [comment]
@VinodAnandan - the upgrade to Quarkus 3.8.5 is handled in this issue #30160, and discussing the general upgrade is not related to this issue. As KC25 has been delayed for other reasons, it is now part of KC25. 

Could someone please check if the issue is now resolved when testing the KC24 release branch or the main branch?

**Comment by vmuzikar** [comment]
> Could someone please check if the issue is now resolved when testing the KC24 release branch or the main branch?

As far as I understand, this is blocked by https://github.com/quarkusio/quarkus/issues/40780, no?

**Comment by gsmet** [comment]
My understanding is that this was solved in 3.8.5.

**Comment by vmuzikar** [comment]
@cwelch-rh Would you be able to test the use case with your providers again when you get a chance?

**Comment by VinodAnandan** [comment]
Thanks @ahus1. Yes, I inquired about the Quarkus update because these regressions and security fixes were blocked on the Quarkus 3.8.5 release, and this issue has been moved to the Keycloak 26.0.0 milestone. I wanted to clarify the deprioritisation and ensure there is no misunderstanding regarding the priority. In my opinion, the regression and security fixes should be prioritised. Is there another issue or discussion explaining the rationale for Keycloak's dependency on Quarkus 3.8.x?

By the way, this issue is still included in the Keycloak 26 milestone. It would be great if it could be included in the correct milestone (Keycloak 25).

**Comment by ahus1** [comment]
@VinodAnandan - discussing the Quarkus 3.8.5 upgrade is off-topic for this issue. It is handled in this issue https://github.com/keycloak/keycloak/issues/30160 as I stated above.

Please stop discussing it here.

**Comment by cwelch-rh** [comment]
@vmuzikar Tested main, looks good to me.

With providers:
>Finished step "io.quarkus.resteasy.reactive.jackson.deployment.processor.ResteasyReactiveJacksonProcessor#handleFieldSecurity" in 0 ms
>Quarkus augmentation completed in 7728ms

Without providers:
>Finished step "io.quarkus.resteasy.reactive.jackson.deployment.processor.ResteasyReactiveJacksonProcessor#handleFieldSecurity" in 0 ms
>Quarkus augmentation completed in 5314ms

**Comment by vmuzikar** [comment]
@cwelch-rh Great to hear that! I'm resolving the issue then.

**Comment by vsun757** [comment]
Are there any timelines for releasing version 24.0.6? I see version 25.0.0 is already released

**Comment by vmuzikar** [comment]
Since Keycloak 25 was already released, 24 will not receive any further patches. 24.0.6 will be available only as Red Hat build of Keycloak.

**Comment by vsun757** [comment]
What does Red Hat build mean?

**Comment by vmuzikar** [comment]
@vsun757 It's a supported distribution of Keycloak made by Red Hat. Please see: https://access.redhat.com/products/red-hat-build-of-keycloak.
